src/garage/envs/multi_task_env.py

```diff
@@ -0,0 +1,139 @@
+        task_args (dict): Arguments to be passed to GarageEnv constructer.
+            Keys should be same as keys of task_env_cls_dict.
+        task_kwargs (dict): Arguments to be passed to GarageEnv constructer.
+            Keys should be same as keys of task_env_cls_dict.
+        sample_strategy (str): The sample strategy to be used when
+            "sample_tasks" is called. Acceptable values are 'all' and 'random'.
+    """
+
+    def __init__(self,
+                 task_env_cls_dict,
+                 task_args,
+                 task_kwargs,
+                 sample_strategy='all'):
+
+        self._task_envs = []
+        self._active_task = 0
+        self._valid_sample_strategies = ['all', 'random']
+
+        if sample_strategy not in self._valid_sample_strategies:
+            raise ValueError('Invalid sample strategy.')
+
+        self._sample_strategy = sample_strategy
+
+        for task, env_cls in task_env_cls_dict.items():
+            env_kwargs = task_kwargs[task]
+            env_args = task_args[task]
+            task_env = env_cls(*env_args, **env_kwargs)
+            if self._task_envs:
+                if (task_env.observation_space.shape !=
+                        self.active_env.observation_space.shape):
+                    raise ValueError(
+                        'Observation space of all tasks should be same.')
+                if (task_env.action_space.shape !=
+                        self.active_env.action_space.shape):
+                    raise ValueError(
+                        'Action space of all tasks should be same.')
+
+            self._task_envs.append(task_env)
+
+    @property
+    def task_space(self):
+        """Task Space."""
+        one_hot_ub = np.ones(self.num_tasks)
+        one_hot_lb = np.zeros(self.num_tasks)
+        return akro.Box(one_hot_lb, one_hot_ub)
+
+    @property
+    def active_task(self):
+        """Id of currently active task."""
+        return self._active_task
+
+    @property
+    def active_env(self):
+        """Environment of currently active task."""
+        return self._task_envs[self.active_task]
+
+    @property
+    def num_tasks(self):
+        """Total number of tasks."""
+        return len(self._task_envs)
+
+    @property
+    def action_space(self):
+        """Task's Action space."""
+        return self.active_env.action_space
+
+    @property
+    def observation_space(self):
+        """Observation space."""
+        return self.active_env.observation_space
+
+    @property
+    def horizon(self):
+        """Horizon of currently active task env."""
+        return self.active_env.horizon
+
+    @property
+    def valid_sample_strategies(self):
+        """Supported sample strategies."""
+        return self._valid_sample_strategies
+
+    def reset(self, **kwargs):
+        """Reset the active task env."""
+        return self.active_env.reset(**kwargs)
+
+    def sample_tasks(self, n_tasks=None):
```

Comment:
okay -- either task sampling is internal or it is external. it doesn't make sense for this to be a public method i sampling is an internal process.

Reply:
@ryanjulian I am kind of confused by what you are saying? Can you elaborate more on this?

Reply:
I'm arguing that either (1) this environment samples tasks itself when it needs them (in which case this method should be private or not exist), or (2) it receives already-sampled collections of environments and this method should not be part of its implementation at all.

If we need an environment sampler, then that's a different object.

It's not clear this proxy object design (where the wrapper holds a bunch of envs inside of it and switches the "active" one) is necessary under KR's new sampler API, but I can see us commiting this version while we wait for the multi-task friendly API to get merged.

Reply:
I looked into the new sampler API in more detail and I see your concern now. 
How about we let `MultiTaskEnv` sample a new task internally after `step` is called `max_path_length` times?

Reply:
If we have an environment sampler then we can directly feed those sampled environments to KR's new sampler API. In that case we don't really need a `MultiTaskEnv`, right?. or am I missing something here?

Reply:
i don't think KR's design includes an environment sampler (that is, a module which chooses environments). @krzentner

Reply:
In a finite number of goals case, implicitly switching them in reset makes sense since we can process them in preprocessor. But for meta-RL algorithms, this is making things hard since the assumption of these algos is within a meta batch, we only sample one task. 

I suggest that we do task setting explicitly in the sampler using API like `sample_tasks`+`set_task`.

Reply:
can you explain the assertion 

> the assumption of these [meta-RL] algos is within a meta batch, we only sample one task

meta-RL algos don't have much meaning on a single task, and I don't know of any that only sample one task in a meta-batch.

Reply:
For example, for maml, we need a batch of data that only contains one task (normally sampled from a continuous space) to compute the inner adaptation.

Reply:
Of course, for one optimization step, there are normally more than one meta batches.

Reply:
How do you define "task" and "meta-batch"?

My definitions are:
* task: A fully-defined MDP. This could be a goal-conditioned environment and a (single) goal, or a multi-task environment and a (single) reward function
* meta-batch: A set of multiple, different MDPs (tasks). Formally, a meta-batch is only the MDPs, but we usually use it to mean a set of samples from these MDPs and sometimes also labels (or contexts, e.g. goals) for the MDPs.

Using these definitions, it seems to me that for MAML training you need meta-batches (batches with data from multiple MDPs and the MDP labels themselves), and for MAML adaptation you really just need one (non-meta) batch from some task of your choice.

I would suggest that in an implementation of MAML, if you need both of these behaviors in a single process, you might want to create two different samplers for simplicity and efficiency. Alternatively, you could keep a single sampler and reconfigure it between the training and adaptation stages.

Reply:
I agree with your definition of task but not the meta-batch. Here is my definition of meta-batch: it is a batch of data that is fed into the algorithms for optimization. Each meta-batch contains data collected with a single task (an MDP) and each optimization step uses K meta-batches to optimize the models. This definition is the one that we are using in meta-world paper and originally MAML implementation/paper